src/jdk.javadoc/share/classes/jdk/javadoc/doclet/Reporter.java

```diff
@@ -25,41 +25,121 @@
+ * and may additionally be associated with an {@link Element element} or a
+ * {@link DocTreePath tree node} in a documentation comment.
+ * Other messages may be written directly to one of two streams that are informally
+ * for use by "standard output" and "diagnostic output", where "standard output"
+ * means the output that is the expected result of executing some operation,
+ * such as the command-line help that is generated when using a {@code --help} option,
+ * and "diagnostic output" refers to any errors, warnings and other output that is
+ * a side-effect of executing the operation.
+ *
+ * <p>The exact manner in which diagnostics are output is unspecified and depends
+ * on the enclosing context. For example:
+ * <ul>
+ * <li>The {@link javax.tools.DocumentationTool} API allows a client to specify a
+ * {@link javax.tools.DiagnosticListener} to which diagnostics will be
+ * {@link javax.tools.DiagnosticListener#report reported}. If no listener is specified,
+ * diagnostics will be written to a given stream, or to {@code System.err} if no such
+ * stream is provided.
+ * <li>The {@link java.util.spi.ToolProvider} API allows a client to specify
+ * the streams to be used for reporting standard and diagnostic output.
+ * </ul>
  *
  * @since 9
  */
 public interface Reporter {
 
     /**
-     * Print error message and increment error count.
+     * Prints a diagnostic message.
+     *
+     * @param kind    the kind of diagnostic
+     * @param message message to print
+     */
+    void print(Diagnostic.Kind kind, String message);
+
+    /**
+     * Prints a diagnostic message related to a tree node in a documentation comment.
+     *
+     * @param kind    the kind of diagnostic
+     * @param path    the path for the tree node
+     * @param message the message to be printed
+     */
+    void print(Diagnostic.Kind kind, DocTreePath path, String message);
+
+    /**
+     * Prints a diagnostic message related to an element.
      *
-     * @param kind specify the diagnostic kind
-     * @param msg message to print
+     * @param kind    the kind of diagnostic
+     * @param element the element
+     * @param message the message to be printed
      */
-    void print(Diagnostic.Kind kind, String msg);
+    void print(Diagnostic.Kind kind, Element element, String message);
 
     /**
-     * Print an error message and increment error count.
+     * Prints a diagnostic message related to a position within a range of characters in a file.
+     * The positions are all 0-based character offsets from the beginning of content of the file.
      *
-     * @param kind specify the diagnostic kind
-     * @param path the DocTreePath of item where the error occurs
-     * @param msg message to print
+     * @param kind    the kind of diagnostic
+     * @param file    the file
+     * @param start   the beginning of the enclosing range
+     * @param pos     the position
```

Comment:
Unless there's a reason not to, I would prefer to have an inequality clearly showing how these 3 indexes relate to each other. For example, is it the case that `{@code start <= pos <= end}`?

Reply:
Will do. I had this in an earlier draft. I backed out because I also wanted to bound `start` and `end` but it is not common in API to indicate indices must be non-negative, and it is an IO operation (and potential exception) to verify the upper bound is valid.